File: packages/types/src/worklog.ts
/**
 * Worklog and Timer types for time tracking
 */

export interface IWorklogUserDetail {
  id: string;
  first_name: string;
  last_name: string;
  avatar: string | null;
  avatar_url: string | null;
  display_name: string;
}

export interface IWorklog {
  id: string;
  issue: string;
  user: string;
  project: string;
  workspace: string;
  duration: number; // minutes
  hours: number;
  minutes: number;
  display_duration: string; // "3h 30m"
  description: string;
  logged_at: string;
  source: "manual" | "timer";
  user_detail: IWorklogUserDetail;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  created_by: string;
  updated_by: string | null;
}

export interface IWorklogCreatePayload {
  hours: number;
  minutes: number;
  description?: string;
}

export interface IWorklogUpdatePayload {
  hours?: number;
  minutes?: number;
  description?: string;
}

export interface IWorklogListResponse {
  results: IWorklog[];
  total_duration: number; // total minutes
  total_display_duration: string; // "7h 35m"
}

export interface IActiveTimer {
  id: string;
  issue: string;
  project: string;
  workspace: string;
  user: string;
  start_time: string;
  elapsed_seconds: number;
  issue_identifier: string; // "PROJ-123"
  created_at: string;
  updated_at: string;
}

export interface ITimerStartPayload {
  issue_id: string;
  project_id: string;
}

export interface ITimerStopPayload {
  description?: string;
}

File: apps/api/plane/api/serializers/worklog.py
# Worklog Serializers for Plane Community Edition
# SPDX-License-Identifier: AGPL-3.0-only

from rest_framework import serializers

from .base import BaseSerializer
from .user import UserLiteSerializer
from plane.db.models import Worklog, ActiveTimer


class WorklogSerializer(BaseSerializer):
    """
    Full serializer for worklog entries.

    Includes user details for display in activity trail.
    """

    user_detail = UserLiteSerializer(source="user", read_only=True)
    display_duration = serializers.CharField(read_only=True)
    hours = serializers.IntegerField(read_only=True)
    minutes = serializers.IntegerField(read_only=True)

    class Meta:
        model = Worklog
        fields = "__all__"
        read_only_fields = [
            "id",
            "created_by",
            "updated_by",
            "created_at",
            "updated_at",
            "workspace",
            "project",
            "deleted_at",
            "user",
            "issue",
        ]

    def validate_duration(self, value):
        """Duration must be at least 1 minute."""
        if value is not None and value < 1:
            raise serializers.ValidationError(
                "Duration must be at least 1 minute."
            )
        return value


class WorklogCreateSerializer(BaseSerializer):
    """
    Serializer for creating manual worklog entries.

    Accepts hours + minutes separately, converts to total minutes.
    """

    hours = serializers.IntegerField(
        required=False, default=0, min_value=0
    )
    minutes = serializers.IntegerField(
        required=False, default=0, min_value=0, max_value=59
    )
    description = serializers.CharField(
        required=False, allow_blank=True, default=""
    )

    class Meta:
        model = Worklog
        fields = ["hours", "minutes", "description", "logged_at"]

    def validate(self, data):
        hours = data.get("hours", 0)
        minutes = data.get("minutes", 0)
        total = (hours * 60) + minutes

        if total < 1:
            raise serializers.ValidationError(
                "Total time must be at least 1 minute."
            )

        data["duration"] = total
        return data


class WorklogLiteSerializer(BaseSerializer):
    """
    Lightweight serializer for worklog entries.

    Used in summary/aggregate responses.
    """

    display_duration = serializers.CharField(read_only=True)

    class Meta:
        model = Worklog
        fields = [
            "id",
            "duration",
            "display_duration",
            "description",
            "source",
            "logged_at",
            "created_at",
        ]
        read_only_fields = fields


class ActiveTimerSerializer(BaseSerializer):
    """
    Serializer for the currently running timer.

    Includes elapsed time calculated from start_time.
    """

    elapsed_seconds = serializers.SerializerMethodField()
    issue_identifier = serializers.SerializerMethodField()

    class Meta:
        model = ActiveTimer
        fields = [
            "id",
            "issue",
            "project",
            "workspace",
            "user",
            "start_time",
            "lease_expires_at",
            "last_heartbeat_at",
            "elapsed_seconds",
            "issue_identifier",
            "created_at",
            "updated_at",
            "deleted_at",
        ]
        read_only_fields = [
            "id",
            "created_by",
            "updated_by",
            "created_at",
            "updated_at",
            "workspace",
            "deleted_at",
            "user",
        ]

    def get_elapsed_seconds(self, obj):
        """Calculate elapsed seconds since timer started."""
        from django.utils import timezone

        if obj.start_time:
            delta = timezone.now() - obj.start_time
            return int(delta.total_seconds())
        return 0

    def get_issue_identifier(self, obj):
        """Get the issue identifier (e.g., PROJ-123) for display."""
        if obj.issue and obj.issue.project:
            return f"{obj.issue.project.identifier}-{obj.issue.sequence_id}"
        return None


class TimerHeartbeatSerializer(serializers.Serializer):
    """
    Serializer for renewing an active timer lease.
    """

    lease_token = serializers.CharField(required=True, allow_blank=False)


class TimerStartSerializer(serializers.Serializer):
    """
    Serializer for starting a timer.

    Only requires issue_id and project_id.
    """

    issue_id = serializers.UUIDField(required=True)
    project_id = serializers.UUIDField(required=True)
